Add ERC1167 library (minimal proxy)

Amxx opened the pull request:
Proposal, to be discussed, for an ERC1167 ProxyFactory.

Fixes #2437

frangio commented on contracts/drafts/Clones.sol:
```diff
@@ -0,0 +1,53 @@
+
+/**
+ * @dev https://eips.ethereum.org/EIPS/eip-1167[EIP 1167] is a standard for
+ * deploying minimal clone contracts.
```
```suggestion
 * deploying minimal proxy contracts, also known as "clones".
```

Following the name of the EIP.

frangio commented on contracts/drafts/Clones.sol:
```diff
@@ -0,0 +1,53 @@
+/**
+ * @dev https://eips.ethereum.org/EIPS/eip-1167[EIP 1167] is a standard for
+ * deploying minimal clone contracts.
+ *
```
Copied for the EIP.

```suggestion
 *
 * > To simply and cheaply clone contract functionality in an immutable way, this standard specifie
 * > a minimal bytecode implementation that delegates all calls to a known, fixed address.
 *
```

frangio commented on contracts/drafts/Clones.sol:
```diff
@@ -0,0 +1,53 @@
+        }
+    }
+
+    function clone2(address master, bytes32 salt) internal returns (address instance) {
```
I would prefer a name like the following:

```suggestion
    function cloneDeterministic(address master, bytes32 salt) internal returns (address instance) {
```

frangio commented on contracts/drafts/Clones.sol:
```diff
@@ -0,0 +1,53 @@
+        require(instance != address(0), "ERC1167: create2 failed");
+    }
+
+    function predict2(address master, bytes32 salt, address deployer) internal pure returns (address predicted) {
```
And here maybe `predictAddress`.

Amxx replied:
What about `predictDeterministicAddress` ?

frangio replied:
Sounds good!

frangio reviewed (commented):
The library looks good, we need some tests for it. The EIP is Final so we can move it to the `proxy` directory.

Amxx commented:
> The library looks good, we need some tests for it. The EIP is Final so we can move it to the `proxy` directory.

I moved this to contracts/proxy and added tests

frangio commented on contracts/proxy/Clones.sol:
```diff
@@ -13,6 +13,11 @@ pragma solidity >=0.6.0 <0.8.0;
+    /**
+     * @dev Deploys and return the address of a clone that mimics the behaviour of `master`.
+     *
+     * This function uses the create opcode, which should never revert.
```
I just realized that this could potentially revert in an out of gas situation. I guess this means we should revert if `instance` is 0.

Amxx replied:
In out of gas, would the instance be set to 0 (which we should check) ? I was under the impression that the call would just revert on its own

frangio commented:
I moved your content around in the docs to keep it either at the top of the proxies page or inside the contract description.

I know that we have reservations with this content organization but let's stick to the current thing until we review it.

frangio reviewed (approved):
I think this looks good!
